networkx/drawing/layout.py

```diff
@@ -1110,6 +1111,214 @@ def multipartite_layout(G, subset_key="subset", align="vertical", scale=1, cente
     return pos
 
 
+def estimate_factor(n, swing, traction, speed, speed_efficiency, jitter_tolerance):
+    """
+    ForceAtlas2 helper function
+    Computes scaling factor for force
+    """
+    import numpy as np
+
+    # estimate jitter
+    opt_jitter = 0.05 * np.sqrt(n)
+    min_jitter = np.sqrt(opt_jitter)
+    max_jitter = 10
+    min_speed_efficiency = 0.05
+
+    other = min(max_jitter, opt_jitter * traction / n**2)
+    jitter = jitter_tolerance * max(min_jitter, other)
+
+    if swing / traction > 2.0:
+        if speed_efficiency > min_speed_efficiency:
+            speed_efficiency *= 0.5
+        jitter = max(jitter, jitter_tolerance)
+    if swing == 0:
+        target_speed = np.inf
+    else:
+        target_speed = jitter * speed_efficiency * traction / swing
+
+    if swing > jitter * traction:
+        if speed_efficiency > min_speed_efficiency:
+            speed_efficiency *= 0.7
+    elif speed < 1000:
+        speed_efficiency *= 1.3
+
+    max_rise = 0.5
+    speed = speed + min(target_speed - speed, max_rise * speed)
+    return speed, speed_efficiency
+
+
+def forceatlas2_layout(
+    G,
+    pos=None,
+    n_iter=100,
+    jitter_tolerance=1.0,
+    scaling_ratio=2.0,
+    gravity=1.0,
+    distributed_action=False,
+    strong_gravity=False,
+    adjust_sizes=False,
+    dissuade_hubs=False,
+    edge_weight_influence=False,
+    linlog=False,
+    dim = 2,
+):
+    """Forceatlas2 layout for networkx
+
+    See [1] for more info on the parameters
+
+    [1]: https://journals.plos.org/plosone/article/file?id=10.1371/journal.pone.0098679&type=printable
+    Parameters
+    ----------
+    G : nx.Graph
+       Netwrorkx graph
+    pos: dict or None
+       Optional starting positions
+    n_iter: int
+        Simulation steps
+    jitter_tolerance : float
+        Jitter  tolerance  for  adjusting  speed  of  layout
+        generation
+    scaling_ratio : float
+        Controls  force scaling  constants k_attraction  and
+        k_repulsion
+    distributed_action : bool
+    strong_gravity : bool
+        Controls the  "pull" to  the center  of mass  of the
+        plot (0,0)
+    adjust_sizes: bool
+        Prevent node overlapping in the layout
+    dissuade_hubs : bool
+        Prevent hub clustering
+    edge_weight_influence : bool
+        Generate layout with or without considering the edge
+        weights
+    linlog : bool
+        Use log attraction rather than linear attraction
+    dim: int,
+       Sets the dimensions of the layout. This parameter is ignored if pos is given.
+
+    Examples
+    --------
+    >>> import networkx as nx
+    >>> G = nx.florentine_families_graph()
+    >>> nx.draw(G, pos = nx.forceatlas2_layout(G))
+    """
+    import numpy as np
+
+    # parse optional pos positions
+    if pos is None:
+        pos = nx.random_layout(G, dim = dim)
+    else:
+        dim = len(next(iter(pos.values())))
```

Comment:
```suggestion
    if pos is None:
        pos = nx.random_layout(G, dim=dim)
    else:
        dim = len(nx.utils.arbitrary_element(pos.values()))
```
There are many ways to get an arbitrary element, so to standardize it we added another one so we can change it uniformly if we want to later.
Style guidelines say keyword args `=` sign doesn't get spaces around it

Reply:
I wasn't aware of this utils, good for pointing it out.